torch/_inductor/scheduler.py

```diff
@@ -1916,6 +1943,145 @@ def create_scheduler_node(self, node: ir.Operation) -> BaseSchedulerNode:
+                input_node=orig_node.input_node,
+                ignore_defer_reduction_split=True,
+                split_hints=split_hints,
+            )
+
+            new_node_name = new_node.realize()
+            new_buffer = V.graph.name_to_buffer[new_node_name]
+            assert isinstance(orig_buffer, ir.ComputedBuffer)  # linter
+            assert isinstance(new_buffer, ir.ComputedBuffer)  # linter
+            self.replace_operation_buffer(orig_buffer, new_buffer)
+
+            # After creating the Reduction node, its layout is fixed by
+            # https://github.com/pytorch/pytorch/blob/main/torch/_inductor/graph.py#L1537
+            # Also call `.as_fixed()` to all new reduction nodes.
+            new_buffer.layout = new_buffer.layout.as_fixed()
+            for i in range(orig_num_nodes, len(V.graph.operations)):
+                middle_buffer = V.graph.operations[i]
+                assert isinstance(middle_buffer, ir.ComputedBuffer)  # linter
+                middle_buffer.layout = middle_buffer.layout.as_fixed()
+
+            # Add the new reduction nodes to the scheduler
+            new_scheduler_node = self.create_scheduler_node(new_buffer)
+            self.nodes[idx] = new_scheduler_node
+            self.name_to_node[snode.get_name()] = new_scheduler_node
+            self.name_to_fused_node[snode.get_name()] = new_scheduler_node
+
+            # It's possible that there is no additional middle-layer nodes
+            if len(V.graph.operations) == orig_num_nodes:
+                return
+
+            # Create SchedulerNodes for middle-layer reduction nodes
+            orig_num_scheduler_node = len(self.nodes)
+            for i in range(orig_num_nodes, len(V.graph.operations)):
+                middle_buffer = V.graph.operations[i]
+                assert isinstance(middle_buffer, ir.ComputedBuffer)  # linter
+                middle_buffer.layout = middle_buffer.layout.as_fixed()
+                middle_node = self.create_scheduler_node(middle_buffer)
+                self.nodes.append(middle_node)
+                self.name_to_node[middle_node.get_name()] = middle_node
+                self.name_to_buf.update(
+                    {buf.get_name(): buf for buf in middle_node.get_outputs()}
+                )
+                self.name_to_fused_node[middle_node.get_name()] = middle_node
+
+            # The node whose users are the orig_node (un-split reduction) should be updated to be the first middle reduction node
+            for node in self.nodes:
+                for buf in node.get_outputs():
+                    for user in buf.users:
+                        if user.node.get_name() == snode.get_name():
+                            user.node = self.nodes[orig_num_scheduler_node]
+            # Update the dependencies (buf.usesrs) of the new reduction nodes
+            self.compute_dependencies(
+                self.nodes[orig_num_scheduler_node:] + [new_scheduler_node]
```

Comment:
In each iteration here we're updating  nodes/dependencies. this might be slow. Maybe it would make sense to collect all the buffers we're going to split first, then split, then update once ?